src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,209 @@
+            //    public string State { get; set; }
+            // }
+            var rootNodeValidationErrors = new Dictionary<Type, List<string>>();
+
+            Validate(modelType, visitedTypes, rootNodeValidationErrors);
+
+            foreach (var validationError in rootNodeValidationErrors)
+            {
+                foreach (var validationErrorMessage in validationError.Value)
+                {
+                    // Add error message to model state as exception to avoid
+                    // disclosing details to end user as SerializableError sanitizes the
+                    // model state errors having exceptions with a generic message when sending
+                    // it to the client.
+                    modelStateDictionary.TryAddModelError(
+                        validationError.Key.FullName,
+                        new InvalidOperationException(validationErrorMessage));
+                }
+            }
+        }
+
+        private void Validate(
+            Type modelType,
+            HashSet<Type> visitedTypes,
+            Dictionary<Type, List<string>> errors)
+        {
+            // Need not handle Key-Value pair specially here (for example, when the model type is Dictionary<,>
+            // which implements IEnumerable<KeyValuePair<TKey, TValue>>) as the model type here would be 
+            // KeyValuePair<TKey, TValue> where TKey and TValue are public properties which would also 
+            // be probed for Required attribute validation.
+            if (modelType.IsGenericType())
+            {
+                var enumerableOfT = modelType.ExtractGenericInterface(typeof(IEnumerable<>));
+                if (enumerableOfT != null)
+                {
+                    modelType = enumerableOfT.GetGenericArguments()[0];
+                }
+            }
+
+            if (ExcludeTypeFromValidation(modelType))
+            {
+                return;
+            }
+
+            // Avoid infinite loop in case of self-referencing properties
+            if (!visitedTypes.Add(modelType))
+            {
+                return;
+            }
+
+            Dictionary<Type, List<string>> cachedErrors;
+            if (_cachedValidationErrors.TryGetValue(modelType, out cachedErrors))
+            {
+                foreach (var validationError in cachedErrors)
+                {
+                    errors.Add(validationError.Key, validationError.Value);
+                }
+
+                return;
+            }
+
+            foreach (var propertyHelper in PropertyHelper.GetProperties(modelType))
+            {
+                var propertyInfo = propertyHelper.Property;
+                var propertyType = propertyInfo.PropertyType;
+
+                // Since DefaultObjectValidator can handle Required attribute validation for reference types,
+                // we only consider value types here.
+                if (propertyType.IsValueType() && !propertyType.IsNullableValueType())
```

Comment:
Actually `IsValueType` and `IsNullableValueType` are our own helper extensions methods which call `TypeInfo`.

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Common/TypeExtensions.cs#L88